File: backend/server/core/sse_protocol.py
"""SSE event protocol (pg_notify -> pubsub -> /api/events/{session_id}).

Each SSE frame is `event: <SseEvent>` with a JSON list of entity ids as
data. Mirrored in frontend/src/lib/sockets/sse-protocol.ts — keep in sync.
"""
from enum import Enum


class SseEvent(str, Enum):
    FLASHCARD = "flashcard"    # data: list[str] — flashcard ids to (re)fetch
    DOCUMENTS = "documents"    # data: list[str] — document ids to (re)fetch

File: frontend/src/lib/sockets/sse-protocol.ts
/**
 * SSE event protocol (/api/events/{sessionId}).
 *
 * Each frame is `event: <SseEvent>` with a JSON list of entity ids as data.
 * Mirrored in backend/server/core/sse_protocol.py — keep the two in sync.
 */

export enum SseEvent {
  Flashcard = 'flashcard',
  Documents = 'documents'
}

/** Both current events carry a list of entity ids to (re)fetch. */
export type SsePayload = string[]

export function parseSseEvent(
  eventName: string,
  rawData: string
): { event: SseEvent; ids: SsePayload } | null {
  if (!Object.values(SseEvent).includes(eventName as SseEvent)) return null

  try {
    const data = JSON.parse(rawData)

    if (!Array.isArray(data)) return null

    return { event: eventName as SseEvent, ids: data.map(String) }
  } catch {
    return null
  }
}
